Fix reconnection issue in cli

haslinghuis opened the pull request:
On master it is not possible to go to `OSD` or `Sensor` tab from `CLI`.

This PR should solve this by using an interval to check for a valid connection and removing the duplicate call to the tab cleanup function by removing the call in the disconnect routine in `$('div.connect_controls a.connect')` and only use the call in the tab change request in `$('a', ui_tabs)`.

EDIT:
After lots of testing and input from @Asizon digging further the issue is sending or disconnecting when device_lost has occurred. Rebasing the serial code did solve the issue and some bugs were fixed.   

Tested on Matek F411 on Linux (Fedora 32) and Matek F722 on Windows (10)

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

Asizon commented:
@haslinghuis behaviour here is a litle weird, sometimes debug shows "CLI closed", "device rebooting", and sometimes it doesnt shows. Also when cli disconect reconect the tab is complety blank and doesnt goes to welcome tab.

haslinghuis commented:
@Asizon Currently it only works with auto-connect enabled. Have it working now locally, needed some additional hacks in serial and msp, but still testing and debugging.

mikeller commented:
Interestingly I cannot reproduce this problem in the first place (on linux) - when in the CLI tab, clicking on either the OSD or the Sensors tab gets me there after the flight controller reboot. Is there a different way to reproduce this?

haslinghuis commented:
It's happening on a Matek F411 on Fedora 32 and Windows 10 both on master with latest firmware.

haslinghuis commented:
This is really weird. I removed all previous code and rebased serial based code. On the same machine switching between two branches in different terminals this still solves the issue.

Asizon reviewed (approved):
Wow!All works as expected now.

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

mikeller commented on src/js/serial.js:
```diff
@@ -429,10 +434,11 @@ const serial = {
 
         if (GUI.connected_to || GUI.connecting_to) {
-            $('a.connect').click();
+            $('a.connect').trigger('click');
```
What is the reason for this change?

haslinghuis replied:
New syntax - doing exactly the same.

mikeller commented on src/js/main.js:
```diff
@@ -262,7 +262,7 @@ function startProcess() {
                     $('a.connect').click();
                 } else {
-                    self.disconnect();
+                    serial.disconnect();
```
Good find!

mikeller commented on src/js/serial.js:
```diff
@@ -231,23 +234,22 @@ const serial = {
-                    checkChromeRuntimeError();
-
+                    if (chrome.runtime.lastError) {
+                        console.log(chrome.runtime.lastError.message);
```
What is the reason that we aren't using `checkChromeRuntimeError()` here?

haslinghuis replied:
Maybe not needed anymore but I should not hurt. Needed the code to have less overhead to improve timing.

blckmn commented:
AUTOMERGE: (PASS)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS
